### ISSUE TITLE
Store ack pruning sequence after completing an upgrade in UpgradeOpen

### ISSUE THREAD

**Issue by DimitrisJim** [cosmos/ibc-go#5413]
## Summary

Parent issue #3937

Basically, [this comment](https://github.com/cosmos/ibc-go/issues/3937#issuecomment-1819033944). Can store the sequence after completing (in `WriteUpgradeOpen`) to be available to use for pruning acknowledgements.

____

#### For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate contributors tagged/assigned

**Comment by colin-axner** [comment]
I think this issue is non-trivial. It looks to me we store the [ack sequence lexographically](https://github.com/cosmos/ibc-go/blob/main/modules/core/24-host/keys.go#L196). So the order is "1, 10, 2, 29, 3" which leads to inefficient iteration. It is worth verifying with a unit test. 

I'm not sure if there's an efficient search solution for this problem? It seems to me you'd need to iterate over all keys. We could keep track off all post successful upgrade sequence numbers. Then the pruning function would prune all acks not in that set. Once all other acks are pruned, we could discard the tracked set. Otherwise we would need to add back the latest sequence send

@AdityaSripal

**Comment by colin-axner** [comment]
>  Couldn't we store min seq as we encounter it?

Came here to say this. You can keep track of the lowest sequence received packet post upgrade (for UNORDERED channels). So your active pruning sequence starts at 1, you continue pruning until you reach: no ack for seq or active pruning seq == minPostUpgradeAckSeq

So, [in upgrade open](https://github.com/cosmos/ibc-go/blob/04-channel-upgrades/modules/core/04-channel/keeper/upgrade.go#L531):
```go
 
if channel.Order == types.ORDERED {
    k.SetPostUpgradePruningSeq(ctx, k.GetNextSeqRecv(ctx, portID, channelID)
} else {
    // non ORDERED channels will update the pruning sequence when new packets arrive
    // account for previous upgrade pruning
    postUpgradeMinSeq, found := k.GetPostUpgradePruningSeq(ctx, portID, channelID)
    if found {
        // previous pruning should have started
        currentPruningSeq, found := k.GetCurrentPruningSeq(ctx, portID, channelID)
        if !found {
            // previous pruning completed, update pruning sequence to previous post upgrade seq
            k.SetCurrentPruningSeq(ctx, portID, channelID, postUpgradeMinSeq)
        }
    }
     
    // reset to indicate no new packets received yet
    k.SetPostUpgradePruningSeq(ctx, 0)
}
```

Then [here](https://github.com/cosmos/ibc-go/blob/04-channel-upgrades/modules/core/04-channel/keeper/packet.go#L219):
```go
// optimization for pruning acknowledgements and receipts after a successful channel upgrade
if seq, found := k.GetPostUpgradePruningSeq(ctx, portID, channelID); found {
    if seq == 0 {
        // first receive post upgrade, seq pruning sequence
        k.SetPostUpgradePruningSeq(ctx, packet.GetSequence())
    } else if packet.GetSequence() < seq {
        // store the lowest sequence of the new set of packets
        k.SetPostUpgradePruningSeq(ctx, packet.GetSequence())
    }
}
```

Feel free to adjust naming as necessary. We could set `packet.GetSequence() - 1` to indicate the last expected ack/receipt to prune, but I think it is fine to store the first sequence which we cannot prune. Maybe we have `GetPostUpgradeMinSeq` or something

In pruning logic:
```go
currentPruningSeq, found := k.GetCurrentPruningSeq(ctx, portID, channelID)
if !found {
    return err
}

// first packet received after successful channel upgrade
postUpgradeMinSeq, found := k.GetPostUpgradeMinSeq(ctx, portID, channelID)
if !found {
    return err
}

for i := 0; i++; i < limit {
    prune := k.HasPacketAcknowledgement(ctx, portID, channelID, currentPruningSeq) && currentPruningSeq < postUpgradeMinSeq
    if prune {
        k.DeletePacketAcknowledgement(ctx, portID, channelID, currentPruningSeq)
        if channel.Order == UNORDERED {
            k.DeletePacketReceipt(ctx, portID, channelID, currentPruningSeq) // haha add this function
        }
    } else {
        // pruning complete
        k.DeleteCurrentPruningSeq(ctx, portID, channelID)
        return 
    }
    
    currentPruningSeq++
}

k.SetCurrentPruningSeq(ctx, currentPruningSeq)
return
```

After writing out the logic, it got a bit complicated, if someone sees a way to simplify things that'd be nice. The main issue I saw was when accounting for multiple upgrades where previous pruning might still be on going or may have completed. When it is ongoing it is no problem since you can simply reset the min sequence seen, but when it completed, you need to ensure the current pruning sequence is set to a sane value

**Comment by colin-axner** [comment]
We should really start storing keys with big endian encodings :smile:

**Comment by colin-axner** [comment]
Lets try to implement the above proposal in the meantime ^

**Comment by srdtrk** [comment]
https://github.com/cosmos/ibc-go/issues/5413#issuecomment-1856766156 and @DimitrisJim. I think there is a `log(n)` solution to this problem. I'm not saying we should necessarily implement it but I'm documenting it in case this is somehow relevant in the future. I'll leave some pseudo code here. The assumptions used in this algoritm are
1. The numbers are sequentially present. (i.e. if both `N<M` are in the storage, then all numbers in between them are also in the storage)
2. The numbers are ordered lexicographically.
3. The storage is a [reverse](https://github.com/cosmos/cosmos-sdk/blob/0c0589813be2eb9dd4a5b809fd84ef04866d73a7/core/store/store.go#L32) iterable prefix store.

The idea is that we can determine the first digit of the maxiumum in a fixed number of operations. Once we determine the first digit, we can determine the rest recursively.

Then our algorithm is as follows: (This can be made more efficient, but it'd be harder to follow. I'm also happy to explain it in a call if people are interested as it may be hard to follow anyway)

```go
func getMax(reverseIter) string {
    x := reverseIter.Next()
    // empty iter
    if x == nil { return "" }
    
    // See this by contradiction.
    // Assume that there is a non-9 digit in x and consider x+1 to see that it'd come before x in the reverseIter.
    if !x.EveryDigitIs9() { return x; }
    
    // so if x = "999" then x' = "99". And if x = 9 then x' = ""
    x' := trimPrefix("9", x)
    
    maxCandidate := "0"
    var maxDigitCandidate string
    for i = 8; i > 0; i-- {
        prefixedReverseIter := reverseIter.RestrictToStartWith(string(i))
        
        // Here, I assume that the string(i) was trimmed from r. This doesn't really matter as we can trim it ourselves if not.
        r := prefixedReverseIter.Next()
        if r == nil { panic("violates assumption 1") }
        switch r {
            case x:
                maxCandidate = number(string(i) + r)
                maxDigit = string(i+1)
                break // out of the for loop
            case x':
                // do nothing
            default: // not all digits are 9, use same contradiction logic to end our function
               return string(i) + r
        }
    }
    
    prefixedReverseIter := reverseIter.RestrictToStartWith(maxDigitCandidate)
    // this max is numerical max, not lexicographic (although the arguments are strings for ease of representation)
    return max(maxCandidate, maxDigitCandidate + getMax(prefixedReverseIter))
}
```

**Comment by charleenfei** [comment]
closed with #5438
